src/radical/pilot/raptor/master.py

```diff
@@ -261,84 +247,112 @@ def _state_cb(self, topic, msg):
-        task['uid']               = uid
-        task['task_sandbox_path'] = sbox
-        task['task_sandbox']      = 'file://localhost/' + sbox
-        task['pilot_sandbox']     = cfg.base
-        task['session_sandbox']   = cfg.base + '/../'
-        task['resource_sandbox']  = cfg.base + '/../../'
-        task['resources']         = {'cpu': td['cpu_processes'] *
-                                            td.get('cpu_threads', 1),
-                                     'gpu': td['gpu_processes']}
-
-        # NOTE: the order of insert / state update relies on that order
-        # being maintained through the component's message push, the update
-        # worker's message receive up to the insertion order into the update
-        # worker's DB bulk op.
-        self._log.debug('insert %s', uid)
-        self.publish(rpc.STATE_PUBSUB, {'cmd': 'insert', 'arg': task})
-
-        self._log.debug('submit %s', uid)
-        self.advance(task, publish=True, push=True)
+        pass
 
+
+    # --------------------------------------------------------------------------
+    #
+    def submit_workers(self, descr: Dict[str, Union[str, int]],
+                             count: int) -> None:
+        '''
+        Submit`count` workers per given `descr`, and pass the queue info as
+        configuration file.  Do *not* wait for the workers to come up - they are
+        expected to register via the control channel.
+
+        The `descr` dict is expected to support the following keys:
+
+          - named_env       : environment to use (same as master usually)
+          - ranks           : number of MPI ranks per worker
+          - cores_per_rank  : int, number of cores per worker rank
+          - gpus_per_rank   : int, number of gpus per worker rank
+          - worker_class    : str, type name of worker class to execute
+          - worker_file     : str, optional if an RP worker class is used
+
+        Note that only one rank (presumably rank 0) should register with the
+        master - the worker ranks are expected to syncronize their ranks as
+        needed.
+        '''
         with self._lock:
-            self._workers[uid] = dict()
-            self._workers[uid]['state'] = 'NEW'
+
+            tasks    = list()
+            base     = os.environ['RP_TASK_ID']
+
+            cfg            = copy.deepcopy(self._cfg)
+            cfg['descr']   = descr
+            cfg['info']    = self._info
+            cfg['ts_addr'] = self._task_service.addr
+
+            for i in range(count):
+
+                uid        = '%s.worker.%04d' % (base, i)
+                cfg['uid'] = uid
+
+                fname = './%s.json' % uid
+                ru.write_json(cfg, fname)
+
+                td = dict()
+                td['named_env']        = descr.get('named_env')
+                td['cpu_processes']    = descr['cpu_processes']
+                td['cpu_process_type'] = 'MPI'
+                td['cpu_thread_type']  = 'POSIX'
```

Comment:
better to use corresponding constants (from `radical.pilot.constants`)

Reply:
will do

Reply:
will do